File: src/pythonnative/native/android/src/main/java/com/pythonnative/runtime/components/PNColor.kt
package com.pythonnative.runtime.components

import org.json.JSONArray
import org.json.JSONObject

/**
 * Color parsing shared by every manager.
 *
 * Accepts `#RGB`, `#RGBA`, `#RRGGBB`, `#AARRGGBB` (the Android/Python
 * convention used by `parse_color_int`), `rgb(r, g, b)`,
 * `rgba(r, g, b, a)` with `a` in `0..1`, CSS named colors, packed
 * integers (`0xRRGGBB` gains full alpha), and `[r, g, b(, a)]` arrays.
 * Results are signed 32-bit ARGB ints suitable for Android APIs.
 */
object PNColor {
    private val named: Map<String, Int> = mapOf(
        "transparent" to 0x00000000,
        "clear" to 0x00000000,
        "black" to 0xFF000000.toInt(),
        "white" to 0xFFFFFFFF.toInt(),
        "red" to 0xFFFF0000.toInt(),
        "green" to 0xFF008000.toInt(),
        "lime" to 0xFF00FF00.toInt(),
        "blue" to 0xFF0000FF.toInt(),
        "yellow" to 0xFFFFFF00.toInt(),
        "cyan" to 0xFF00FFFF.toInt(),
        "aqua" to 0xFF00FFFF.toInt(),
        "magenta" to 0xFFFF00FF.toInt(),
        "fuchsia" to 0xFFFF00FF.toInt(),
        "gray" to 0xFF808080.toInt(),
        "grey" to 0xFF808080.toInt(),
        "darkgray" to 0xFFA9A9A9.toInt(),
        "darkgrey" to 0xFFA9A9A9.toInt(),
        "lightgray" to 0xFFD3D3D3.toInt(),
        "lightgrey" to 0xFFD3D3D3.toInt(),
        "dimgray" to 0xFF696969.toInt(),
        "dimgrey" to 0xFF696969.toInt(),
        "silver" to 0xFFC0C0C0.toInt(),
        "maroon" to 0xFF800000.toInt(),
        "olive" to 0xFF808000.toInt(),
        "navy" to 0xFF000080.toInt(),
        "teal" to 0xFF008080.toInt(),
        "purple" to 0xFF800080.toInt(),
        "orange" to 0xFFFFA500.toInt(),
        "pink" to 0xFFFFC0CB.toInt(),
        "brown" to 0xFFA52A2A.toInt(),
        "gold" to 0xFFFFD700.toInt(),
        "indigo" to 0xFF4B0082.toInt(),
        "violet" to 0xFFEE82EE.toInt(),
        "coral" to 0xFFFF7F50.toInt(),
        "salmon" to 0xFFFA8072.toInt(),
        "tomato" to 0xFFFF6347.toInt(),
        "crimson" to 0xFFDC143C.toInt(),
        "khaki" to 0xFFF0E68C.toInt(),
        "beige" to 0xFFF5F5DC.toInt(),
        "ivory" to 0xFFFFFFF0.toInt(),
        "tan" to 0xFFD2B48C.toInt(),
        "turquoise" to 0xFF40E0D0.toInt(),
        "skyblue" to 0xFF87CEEB.toInt(),
        "steelblue" to 0xFF4682B4.toInt(),
        "royalblue" to 0xFF4169E1.toInt(),
        "dodgerblue" to 0xFF1E90FF.toInt(),
        "slategray" to 0xFF708090.toInt(),
        "slategrey" to 0xFF708090.toInt(),
        "whitesmoke" to 0xFFF5F5F5.toInt(),
        "snow" to 0xFFFFFAFA.toInt(),
        "seagreen" to 0xFF2E8B57.toInt(),
        "forestgreen" to 0xFF228B22.toInt(),
        "darkgreen" to 0xFF006400.toInt(),
        "lightblue" to 0xFFADD8E6.toInt(),
        "lightgreen" to 0xFF90EE90.toInt(),
        "darkblue" to 0xFF00008B.toInt(),
        "darkred" to 0xFF8B0000.toInt(),
        "orangered" to 0xFFFF4500.toInt(),
        "hotpink" to 0xFFFF69B4.toInt(),
        "deeppink" to 0xFFFF1493.toInt(),
        "chocolate" to 0xFFD2691E.toInt(),
        "plum" to 0xFFDDA0DD.toInt(),
        "lavender" to 0xFFE6E6FA.toInt(),
        "mintcream" to 0xFFF5FFFA.toInt(),
        "gainsboro" to 0xFFDCDCDC.toInt(),
    )

    /** Parse `value` into a signed ARGB int, or `null` when unparseable. */
    fun parse(value: Any?): Int? {
        return when (value) {
            null, JSONObject.NULL -> null
            is Boolean -> null
            is Int -> if (value ushr 24 == 0) value or 0xFF000000.toInt() else value
            is Long -> parseLong(value)
            is Number -> {
                val d = value.toDouble()
                if (d.isNaN() || d.isInfinite()) null else parseLong(d.toLong())
            }
            is String -> parseString(value)
            is JSONArray -> parseArray((0 until value.length()).map { value.opt(it) })
            is List<*> -> parseArray(value)
            else -> null
        }
    }

    /** Parse `value`, falling back to `default` when unparseable. */
    fun parseOr(value: Any?, default: Int): Int = parse(value) ?: default

    /** Replace the alpha channel of `color` with `alpha` in `0..1`. */
    fun withAlpha(color: Int, alpha: Double): Int {
        val a = (alpha.coerceIn(0.0, 1.0) * 255.0).toInt()
        return (color and 0x00FFFFFF) or (a shl 24)
    }

    private fun parseLong(raw: Long): Int {
        val masked = raw and 0xFFFFFFFFL
        val withAlpha = if (masked ushr 24 == 0L) masked or 0xFF000000L else masked
        return withAlpha.toInt()
    }

    private fun parseArray(items: List<Any?>): Int? {
        if (items.size < 3) return null
        val r = channel(items[0]) ?: return null
        val g = channel(items[1]) ?: return null
        val b = channel(items[2]) ?: return null
        val a = if (items.size >= 4) alphaChannel(items[3]) ?: return null else 255
        return argb(a, r, g, b)
    }

    private fun parseString(raw: String): Int? {
        val s = raw.trim()
        if (s.isEmpty()) return null
        if (s.startsWith("#")) return parseHex(s.substring(1))
        val lower = s.lowercase()
        named[lower]?.let { return it }
        if (lower.startsWith("rgb")) return parseFunctional(lower)
        if (lower.startsWith("0x")) return s.substring(2).toLongOrNull(16)?.let { parseLong(it) }
        return s.toLongOrNull(16)?.takeIf { s.length == 6 || s.length == 8 }?.let {
            parseHexDigits(s)
        }
    }

    private fun parseHex(digits: String): Int? {
        if (digits.isEmpty() || !digits.all { it.isLetterOrDigit() }) return null
        val expanded = when (digits.length) {
            3 -> "FF" + digits.map { "$it$it" }.joinToString("")
            4 -> {
                // CSS #RGBA: alpha is the last digit.
                val rgb = digits.substring(0, 3).map { "$it$it" }.joinToString("")
                val a = "${digits[3]}${digits[3]}"
                a + rgb
            }
            6 -> "FF$digits"
            8 -> digits
            else -> return null
        }
        return expanded.toLongOrNull(16)?.toInt()
    }

    private fun parseHexDigits(digits: String): Int? = parseHex(digits)

    private fun parseFunctional(lower: String): Int? {
        val open = lower.indexOf('(')
        val close = lower.lastIndexOf(')')
        if (open < 0 || close < open) return null
        val parts = lower.substring(open + 1, close).split(',', '/', ' ').map { it.trim() }.filter { it.isNotEmpty() }
        if (parts.size < 3) return null
        val r = channel(parts[0]) ?: return null
        val g = channel(parts[1]) ?: return null
        val b = channel(parts[2]) ?: return null
        val a = if (parts.size >= 4) alphaChannel(parts[3]) ?: return null else 255
        return argb(a, r, g, b)
    }

    private fun channel(value: Any?): Int? {
        val d = when (value) {
            is Number -> value.toDouble()
            is String -> {
                val t = value.trim()
                if (t.endsWith("%")) (t.dropLast(1).toDoubleOrNull() ?: return null) * 2.55
                else t.toDoubleOrNull() ?: return null
            }
            else -> return null
        }
        return d.toInt().coerceIn(0, 255)
    }

    private fun alphaChannel(value: Any?): Int? {
        val d = when (value) {
            is Number -> value.toDouble()
            is String -> {
                val t = value.trim()
                if (t.endsWith("%")) (t.dropLast(1).toDoubleOrNull() ?: return null) / 100.0
                else t.toDoubleOrNull() ?: return null
            }
            else -> return null
        }
        // Values above 1 are taken as 0..255 bytes.
        val a = if (d > 1.0) d / 255.0 else d
        return (a.coerceIn(0.0, 1.0) * 255.0 + 0.5).toInt()
    }

    private fun argb(a: Int, r: Int, g: Int, b: Int): Int =
        (a shl 24) or (r shl 16) or (g shl 8) or b
}

File: src/pythonnative/native/ios/Sources/PythonNativeKit/Components/PNColor.swift
import UIKit

/// Color parsing shared by every manager.
///
/// Accepts `#RGB`, `#RGBA`, `#RRGGBB`, `#AARRGGBB` (the PythonNative
/// convention: alpha first), `rgb(r, g, b)`, `rgba(r, g, b, a)`, CSS
/// named colors, `"transparent"`, raw ARGB integers, and
/// `{"light": c, "dark": c}` dictionaries for dynamic colors.
public enum PNColor {
    /// Parse `value` into a `UIColor`, or `nil` when it isn't a color.
    public static func parse(_ value: Any?) -> UIColor? {
        switch value {
        case nil, is NSNull:
            return nil
        case let dict as [String: Any]:
            return dynamic(dict)
        case let number as NSNumber where !(value is Bool):
            return fromARGB(number.int64Value)
        case let int as Int:
            return fromARGB(Int64(int))
        case let string as String:
            return parse(string: string)
        default:
            return nil
        }
    }

    /// Parse a color string; see `parse(_:)` for the accepted grammar.
    public static func parse(string raw: String) -> UIColor? {
        let text = raw.trimmingCharacters(in: .whitespacesAndNewlines)
        if text.isEmpty { return nil }
        let lower = text.lowercased()
        if lower == "transparent" || lower == "clear" || lower == "none" {
            return UIColor.clear
        }
        if text.hasPrefix("#") {
            return parseHex(String(text.dropFirst()))
        }
        if lower.hasPrefix("rgb") {
            return parseRGBFunction(lower)
        }
        if let named = PNNamedColors.table[lower] {
            return fromARGB(named)
        }
        if let hex = parseHex(text), text.count == 6 || text.count == 8 {
            return hex
        }
        return nil
    }

    /// Components (0...1) of `color` in the sRGB space.
    public static func components(_ color: UIColor) -> (r: CGFloat, g: CGFloat, b: CGFloat, a: CGFloat) {
        var r: CGFloat = 0, g: CGFloat = 0, b: CGFloat = 0, a: CGFloat = 0
        if color.getRed(&r, green: &g, blue: &b, alpha: &a) {
            return (r, g, b, a)
        }
        var white: CGFloat = 0
        if color.getWhite(&white, alpha: &a) {
            return (white, white, white, a)
        }
        return (0, 0, 0, 1)
    }

    /// `#AARRGGBB` string for `color`, matching the Python interpolation output.
    public static func hexString(_ color: UIColor) -> String {
        let c = components(color)
        func byte(_ v: CGFloat) -> Int { Int((max(0, min(1, v)) * 255).rounded()) }
        return String(format: "#%02X%02X%02X%02X", byte(c.a), byte(c.r), byte(c.g), byte(c.b))
    }

    /// Build a color from a 32-bit ARGB integer (signed or unsigned).
    public static func fromARGB(_ value: Int64) -> UIColor {
        var argb = value
        if argb < 0 { argb += 0x1_0000_0000 }
        let a = CGFloat((argb >> 24) & 0xFF) / 255
        let r = CGFloat((argb >> 16) & 0xFF) / 255
        let g = CGFloat((argb >> 8) & 0xFF) / 255
        let b = CGFloat(argb & 0xFF) / 255
        return UIColor(red: r, green: g, blue: b, alpha: a)
    }

    // MARK: - Private

    private static func parseHex(_ digits: String) -> UIColor? {
        var hex = digits
        switch hex.count {
        case 3:
            hex = "FF" + hex.map { "\($0)\($0)" }.joined()
        case 4:
            // #RGBA short form (CSS): expand then move alpha to the front.
            let expanded = hex.map { "\($0)\($0)" }
            hex = expanded[3] + expanded[0] + expanded[1] + expanded[2]
        case 6:
            hex = "FF" + hex
        case 8:
            break
        default:
            return nil
        }
        guard let value = Int64(hex, radix: 16) else { return nil }
        return fromARGB(value)
    }

    private static func parseRGBFunction(_ text: String) -> UIColor? {
        guard let open = text.firstIndex(of: "("), let close = text.lastIndex(of: ")"), open < close else { return nil }
        let inner = text[text.index(after: open)..<close]
        let parts = inner.split(whereSeparator: { $0 == "," || $0 == " " || $0 == "/" })
            .map { $0.trimmingCharacters(in: .whitespaces) }
            .filter { !$0.isEmpty }
        guard parts.count == 3 || parts.count == 4 else { return nil }
        func channel(_ s: String) -> CGFloat? {
            if s.hasSuffix("%"), let v = Double(s.dropLast()) { return CGFloat(v / 100) }
            guard let v = Double(s) else { return nil }
            return CGFloat(v / 255)
        }
        func alpha(_ s: String) -> CGFloat? {
            if s.hasSuffix("%"), let v = Double(s.dropLast()) { return CGFloat(v / 100) }
            guard let v = Double(s) else { return nil }
            return CGFloat(v)
        }
        guard let r = channel(parts[0]), let g = channel(parts[1]), let b = channel(parts[2]) else { return nil }
        let a = parts.count == 4 ? (alpha(parts[3]) ?? 1) : 1
        return UIColor(red: clamp(r), green: clamp(g), blue: clamp(b), alpha: clamp(a))
    }

    private static func dynamic(_ dict: [String: Any]) -> UIColor? {
        let light = parse(dict["light"] ?? dict["default"])
        let dark = parse(dict["dark"])
        guard light != nil || dark != nil else { return nil }
        return UIColor { traits in
            if traits.userInterfaceStyle == .dark {
                return dark ?? light ?? .clear
            }
            return light ?? dark ?? .clear
        }
    }

    private static func clamp(_ v: CGFloat) -> CGFloat { max(0, min(1, v)) }
}

/// CSS named colors as ARGB integers.
enum PNNamedColors {
    static let table: [String: Int64] = [
        "black": 0xFF00_0000, "white": 0xFFFF_FFFF, "red": 0xFFFF_0000, "green": 0xFF00_8000,
        "blue": 0xFF00_00FF, "yellow": 0xFFFF_FF00, "cyan": 0xFF00_FFFF, "aqua": 0xFF00_FFFF,
        "magenta": 0xFFFF_00FF, "fuchsia": 0xFFFF_00FF, "gray": 0xFF80_8080, "grey": 0xFF80_8080,
        "silver": 0xFFC0_C0C0, "maroon": 0xFF80_0000, "olive": 0xFF80_8000, "lime": 0xFF00_FF00,
        "teal": 0xFF00_8080, "navy": 0xFF00_0080, "purple": 0xFF80_0080, "orange": 0xFFFF_A500,
        "pink": 0xFFFF_C0CB, "brown": 0xFFA5_2A2A, "gold": 0xFFFF_D700, "indigo": 0xFF4B_0082,
        "violet": 0xFFEE_82EE, "tomato": 0xFFFF_6347, "coral": 0xFFFF_7F50, "salmon": 0xFFFA_8072,
        "crimson": 0xFFDC_143C, "khaki": 0xFFF0_E68C, "beige": 0xFFF5_F5DC, "ivory": 0xFFFF_FFF0,
        "lavender": 0xFFE6_E6FA, "plum": 0xFFDD_A0DD, "orchid": 0xFFDA_70D6, "turquoise": 0xFF40_E0D0,
        "tan": 0xFFD2_B48C, "chocolate": 0xFFD2_691E, "firebrick": 0xFFB2_2222, "darkred": 0xFF8B_0000,
        "darkgreen": 0xFF00_6400, "darkblue": 0xFF00_008B, "darkgray": 0xFFA9_A9A9, "darkgrey": 0xFFA9_A9A9,
        "dimgray": 0xFF69_6969, "dimgrey": 0xFF69_6969, "lightgray": 0xFFD3_D3D3, "lightgrey": 0xFFD3_D3D3,
        "lightblue": 0xFFAD_D8E6, "lightgreen": 0xFF90_EE90, "lightyellow": 0xFFFF_FFE0, "lightpink": 0xFFFF_B6C1,
        "skyblue": 0xFF87_CEEB, "steelblue": 0xFF46_82B4, "royalblue": 0xFF41_69E1, "dodgerblue": 0xFF1E_90FF,
        "deepskyblue": 0xFF00_BFFF, "slategray": 0xFF70_8090, "slategrey": 0xFF70_8090, "whitesmoke": 0xFFF5_F5F5,
        "gainsboro": 0xFFDC_DCDC, "snow": 0xFFFF_FAFA, "seagreen": 0xFF2E_8B57, "forestgreen": 0xFF22_8B22,
        "limegreen": 0xFF32_CD32, "springgreen": 0xFF00_FF7F, "mintcream": 0xFFF5_FFFA, "aquamarine": 0xFF7F_FFD4,
        "hotpink": 0xFFFF_69B4, "deeppink": 0xFFFF_1493, "orangered": 0xFFFF_4500, "darkorange": 0xFFFF_8C00,
        "goldenrod": 0xFFDA_A520, "sienna": 0xFFA0_522D, "peru": 0xFFCD_853F, "wheat": 0xFFF5_DEB3,
        "linen": 0xFFFA_F0E6, "azure": 0xFFF0_FFFF, "honeydew": 0xFFF0_FFF0, "midnightblue": 0xFF19_1970,
        "rebeccapurple": 0xFF66_3399, "mediumpurple": 0xFF93_70DB, "slateblue": 0xFF6A_5ACD, "cadetblue": 0xFF5F_9EA0,
        "cornflowerblue": 0xFF64_95ED, "darkcyan": 0xFF00_8B8B, "darkmagenta": 0xFF8B_008B, "darkviolet": 0xFF94_00D3,
        "darkslategray": 0xFF2F_4F4F, "darkslategrey": 0xFF2F_4F4F, "lightcoral": 0xFFF0_8080, "lightsalmon": 0xFFFF_A07A,
        "lightseagreen": 0xFF20_B2AA, "lightskyblue": 0xFF87_CEFA, "lightslategray": 0xFF77_8899, "lightslategrey": 0xFF77_8899,
        "lightsteelblue": 0xFFB0_C4DE, "mediumblue": 0xFF00_00CD, "mediumseagreen": 0xFF3C_B371, "mediumvioletred": 0xFFC7_1585,
        "olivedrab": 0xFF6B_8E23, "palegreen": 0xFF98_FB98, "paleturquoise": 0xFFAF_EEEE, "palevioletred": 0xFFDB_7093,
        "powderblue": 0xFFB0_E0E6, "rosybrown": 0xFFBC_8F8F, "saddlebrown": 0xFF8B_4513, "sandybrown": 0xFFF4_A460,
        "thistle": 0xFFD8_BFD8, "yellowgreen": 0xFF9A_CD32, "chartreuse": 0xFF7F_FF00, "greenyellow": 0xFFAD_FF2F,
    ]
}
